rhino/src/main/java/org/mozilla/javascript/ThreadSafeCompoundOperationMap.java

```diff
@@ -0,0 +1,101 @@
+package org.mozilla.javascript;
+
+import java.util.Iterator;
+
+class ThreadSafeCompoundOperationMap extends CompoundOperationMap {
+    private boolean closed = false;
```

Comment:
My first reaction is that relying on "closed" to prevent multiple closes might not do what we need unless "closed" is volatile. Have you done any additional analysis on that regarding the threading model. (And if no locking / no volatile is an option we should also document that in comments.)

Reply:
I added the following comment to that class
```
/**
 * This class represents a compound operation performed on a thread safe slot map. As each compound
 * operation creates a new instance the class itself does not need to consider access by multiple
 * threads. This means that the instance fields do not need to be volatile as we are only
 * considering access from this thread.
 */
 ```